frontend/src/metabase/lib/data_grid.js

```diff
@@ -12,9 +12,9 @@ export function pivot(data) {
     // TODO: we assume dimensions are in the first 2 columns, which is less than ideal
     var pivotCol = 0,
         normalCol = 1,
-        cellCol = 2,
-        pivotColValues = distinctValues(data, pivotCol),
+        cellCol = 2, pivotColValues = distinctValues(data, pivotCol),
         normalColValues = distinctValues(data, normalCol);
+
     if (normalColValues.length <= pivotColValues.length) {
```

Comment:
This function is long and has (unnecessary?) mutation which makes it hard to follow. For instance, I'm unable to decipher quickly what `normalColValues.length <= pivotColValues.length` condition means here.

Because this PR adds a remarkable amount of extra complexity, would be good to refactor parts of logic to smaller functions before a merge.